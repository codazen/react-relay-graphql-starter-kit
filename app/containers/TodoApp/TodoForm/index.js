/* flow */

import React from 'react';

type Props = {
  addTodo: ?Function, // callback to create new todo in parent state
};

type State = {
  content: string, // content for new todo
};

/**
 * Form for creating a new todo item
 * Parent component: TodoApp
 * Child components: N/A
 */
export default class TodoForm extends React.Component {

  static defaultProps: Props;

  constructor(props: Props) {
    super(props);
    this.state = {
      content: '',
    };

    (this:any).handleChange = this.handleChange.bind(this);
    (this:any).handleSubmit = this.handleSubmit.bind(this);
  }

  state: State;

  props: Props;

  handleChange(event) {
    const newState = {};
    newState[event.target.name] = event.target.value;
    this.setState(newState);
  }

  handleSubmit(event) {
    const {
      addTodo,
    } = this.props;

    if (addTodo) {
      addTodo(this.state.content);
    }
    event.preventDefault();
    this.setState({
      content: '',
    });
  }

  render() {
    return (
      <div>
        <h2>Todo Form</h2>
        <input
          type='text'
          value={this.state.content}
          name='content'
          onChange={this.handleChange}
        />
        <button onClick={this.handleSubmit}>
          + Todo
        </button>
      </div>
    );
  }
}

TodoForm.defaultProps = {
  addTodo: () => {},
};
