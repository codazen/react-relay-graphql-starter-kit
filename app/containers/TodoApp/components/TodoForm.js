/* @flow */

import React from 'react';

type Props = {
  addTodo: ?Function, // callback to create new todo in parent state
};

type State = {
  content: string, // content for new todo
};

/**
 * Form for creating a new todo item
 * Parent component: TodoApp
 * Child components: N/A
 */
export default class TodoForm extends React.Component {

  constructor(props: Props) {
    super(props);
    this.state = {
      content: '',
    };
  }

  state: State;

  props: Props;

  handleChange = (event: any) => {
    const newState = {};
    newState[event.target.name] = event.target.value;
    this.setState(newState);
  }

  /**
   * Makes call to addTodo mutation
   */
  handleSubmit = (event: any) => {
    const {
      addTodo,
    } = this.props;

    if (addTodo) {
      addTodo(this.state.content);
    }
    event.preventDefault();
    this.setState({
      content: '',
    });
  }

  render() {
    return (
      <div className="todo-form-component">
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            value={this.state.content}
            name="content"
            placeholder="Add a task"
            onChange={this.handleChange}
          />
          <button action="submit">
            +
          </button>
        </form>
      </div>
    );
  }
}
