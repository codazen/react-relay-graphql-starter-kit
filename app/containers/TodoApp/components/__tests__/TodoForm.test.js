import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import TodoForm from '../TodoForm';

// setup React as a global variable
global.React = React;

describe('TodoForm component', () => {
  let todoForm;
  let event;

  // mock addTodo function
  const addTodo = jest.fn();

  describe('Tests with undefined addTodo prop', () => {
    beforeEach(() => {
      todoForm = shallow(<TodoForm />);
      addTodo.mockClear();
    });

    it('Renders correctly', () => {
      expect(shallowToJson(todoForm)).toMatchSnapshot();
    });

    describe('Unit test: handleChange', () => {
      beforeEach(() => {
        event = {
          target: {
            name: 'content',
            value: 'testValue',
          },
        };
      });

      it('handles input change', () => {
        expect(todoForm.instance().state.content).toBe('');
        todoForm.instance().handleChange(event);
        expect(todoForm.instance().state.content).toBe('testValue');
      });
    });

    describe('Unit test: handleSubmit', () => {
      beforeEach(() => {
        event = {
          preventDefault: () => {},
        };
      });

      it('handles submit without addTodo prop', () => {
        todoForm.instance().handleSubmit(event);
        expect(addTodo).not.toHaveBeenCalled();
        expect(todoForm.instance().state.content).toBe('');
      });
    });

    describe('addTodo functionality', () => {
      it('doesn\'t call addTodo on form submit', () => {
        todoForm.find('form').simulate('submit', { preventDefault: () => {} });
        expect(addTodo).not.toHaveBeenCalled();
        expect(todoForm.instance().state.content).toBe('');
      });
    });
  });

  describe('Tests with addTodo prop', () => {
    beforeEach(() => {
      todoForm = shallow(<TodoForm addTodo={addTodo} />);
      addTodo.mockClear();
    });

    it('Renders correctly', () => {
      expect(shallowToJson(todoForm)).toMatchSnapshot();
    });

    describe('Unit test: handleChange', () => {
      beforeEach(() => {
        event = {
          target: {
            name: 'content',
            value: 'testValue',
          },
        };
      });

      it('handles input change', () => {
        expect(todoForm.instance().state.content).toBe('');
        todoForm.instance().handleChange(event);
        expect(todoForm.instance().state.content).toBe('testValue');
      });
    });

    describe('Unit test: handleSubmit', () => {
      beforeEach(() => {
        event = {
          preventDefault: () => {},
        };
      });

      it('handles submit with addTodo prop', () => {
        todoForm.instance().state.content = 'testValue';
        todoForm.instance().handleSubmit(event);
        expect(addTodo).toHaveBeenCalledWith('testValue');
        expect(todoForm.instance().state.content).toBe('');
      });
    });

    describe('addTodo functionality', () => {
      it('calls addTodo on form submit', () => {
        todoForm.find('form').simulate('submit', { preventDefault: () => {} });
        expect(addTodo).toHaveBeenCalled();
        expect(todoForm.instance().state.content).toBe('');
      });
    });
  });
});
