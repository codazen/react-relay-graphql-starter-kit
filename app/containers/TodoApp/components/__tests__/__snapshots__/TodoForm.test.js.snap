exports[`TodoForm component Tests with addTodo prop Renders correctly 1`] = `
<div
  className="todo-form-component">
  <form
    onSubmit={[Function]}>
    <input
      name="content"
      onChange={[Function]}
      placeholder="Add a task"
      type="text"
      value="" />
    <button
      action="submit">
      +
    </button>
  </form>
</div>
`;

exports[`TodoForm component Tests with undefined addTodo prop Renders correctly 1`] = `
<div
  className="todo-form-component">
  <form
    onSubmit={[Function]}>
    <input
      name="content"
      onChange={[Function]}
      placeholder="Add a task"
      type="text"
      value="" />
    <button
      action="submit">
      +
    </button>
  </form>
</div>
`;
